### ISSUE TITLE
storage_class_name = "" is treated as nil

### ISSUE THREAD

**Issue by dpkirchner** [hashicorp/terraform-provider-kubernetes#1379]
### Terraform Version, Provider Version and Kubernetes Version

```
Terraform version: Terraform v1.0.5 on darwin_amd64
Kubernetes provider version: 2.4.1
Kubernetes version: v1.19.9-gke.1900
```

### Affected Resource(s)

- kubernetes_persistent_volume
- kubernetes_persistent_volume_claim

### Terraform Configuration Files
```hcl
resource "kubernetes_persistent_volume" "tmp" {
  metadata {
    name = "tmp"
  }
  spec {
    capacity = {
      storage = "10Gi"
    }

    access_modes       = ["ReadWriteMany"]
    storage_class_name = ""

    persistent_volume_source {
      nfs {
        path   = "/whatever"
        server = var.filestore_nfs_server_ip
      }
    }
  }
}

resource "kubernetes_persistent_volume_claim" "tmp" {
  metadata {
    name      = "tmp"
    namespace = "default"
  }

  spec {
    access_modes       = ["ReadWriteMany"]
    storage_class_name = ""
    volume_name        = kubernetes_persistent_volume.tmp.metadata.0.name

    resources {
      requests = {
        storage = "10Gi"
      }
    }
  }
}

```

### Debug Output

Can't paste the whole thing, contains proprietary information. Here's the relevant bits though: https://gist.github.com/dpkirchner/76447ded8e97ae7286a83faf034420df

### Steps to Reproduce
1. `terraform apply`
2. `kubectl get events -w`

### Expected Behavior
The PVC should have been created with an empty string for the storage class name.

### Actual Behavior
The PVC ends up being created with the default ("standard") storage class name.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by dpkirchner** [comment]
You can work around this by creating a StorageClass with no parameters and specify it in the `storage_class_name` field for the PV and PVC, but I don't know if this is supported. It seems pretty dirty.

**Comment by Akaame** [comment]
Hello, 

I was able to reproduce your issue and even though I do not have any working knowledge of this library, here are the possible culprits

1. https://github.com/hashicorp/terraform-provider-kubernetes/blob/b5770d4b51a1e1314c0f64b9f2c92cac9be7021a/kubernetes/structure_persistent_volume_claim.go#L82
2.https://github.com/hashicorp/terraform-provider-kubernetes/blob/b5770d4b51a1e1314c0f64b9f2c92cac9be7021a/kubernetes/structure_persistent_volume_spec.go#L388

This here is tricky.

https://github.com/hashicorp/terraform-provider-kubernetes/blob/b5770d4b51a1e1314c0f64b9f2c92cac9be7021a/kubernetes/structure_persistent_volume_spec.go#L1230

This checks if the old string is empty and emits addop or replaceop accordingly. However, the zero-state for `storageClass` is not `empty string` ( as it has a special meaning ) but nil which, I think, unearths another can of worms:

```go
// PVC spec
// Name of the StorageClass required by the claim.
// More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#class-1
// +optional
StorageClassName *string `json:"storageClassName,omitempty" protobuf:"bytes,5,opt,name=storageClassName"`
```

```go
// PV spec
// Name of StorageClass to which this persistent volume belongs. Empty value
// means that this volume does not belong to any StorageClass.
// +optional
StorageClassName string `json:"storageClassName,omitempty" protobuf:"bytes,6,opt,name=storageClassName"`
```

I have so several questions here:

1. Why is `storageClassName` a `*string` in one spec and `string` in another?
2. Why is there an omitempty in json encoding tag even though this field being empty has a special meaning?
3. Why is Dynamic Provision suppression bound to storageClassName being empty even though empty/nil strings are known to be a pain point in interoperability?

**Comment by steveizzle** [comment]
To emphasize the validity of this bug here is a reference to the [official kubernetes docs ](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#reserving-a-persistentvolume), where you can see the  comment
> Empty string must be explicitly set otherwise default StorageClass will be set

**Comment by Lillecarl** [comment]
This makes TF unable to handle static configured PV's for the EFS CSI controller too.

**Comment by sarg3nt** [comment]
This is an OLD problem that keeps getting kicked down the road.
https://github.com/hashicorp/terraform-provider-kubernetes/pull/590
https://github.com/hashicorp/terraform-provider-kubernetes/issues/567
and others.
I wouldn't hold out much hope.

**Comment by dinibu** [comment]
Yup, actually any class-less PVCs/PVs can't be described as resources currently. 

I raised this as a core terraform bug, but they redirected to the provider. 

Here's the relevant part:

I'm trying to specify a `kubernetes_persistent_volume_claim` to a manually defined (i.e. not automatically provisioned) Kubernetes persistent volume. Part of this requires that the storage_class_name attribute should be to be an empty string, as per the [K8S docs](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#class-1):


> A PVC with its `storageClassName` set equal to `""` is always interpreted to be requesting a PV with no class, so it can only be bound to PVs with no class (no annotation or one set equal to `""`). A PVC with no `storageClassName` is not quite the same and is treated differently by the cluster, depending on whether the [DefaultStorageClass admission plugin](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#defaultstorageclass) is turned on.


However, it looks like Terraform K8S provider interprets that as "use the default", which in this case, means that it looks up and uses the default storage class as the field value, and so it does not match with a class-less PV as expected.

A literal empty string is required to tell Kubernetes that it should match against class-less persistent volumes. 

It'd be awesome to have a way to set the `storage_class_name` to be an explicit empty string and disable the defaulting behavior from the provider.

**Comment by Lillecarl** [comment]
@dinibu https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/manifest Requires kubernetes access during planning, other than that it'll work just fine. You can convert existing specs with [k2tf](https://github.com/sl1pm4t/k2tf).

**Comment by Lillecarl** [comment]
still broken

**Comment by mike-sol** [comment]
Still broken. Trying to work around it by providing a kubernetes_manifest directly...

**Comment by mike-sol** [comment]
Yes, something like this works to get around this for now:
```hcl
resource "kubernetes_manifest" "persistentvolumeclaim" {

  manifest = yamldecode(<<-EOF
    apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      name: whatever-pvc-name
      namespace: whatever
    spec:
      accessModes:
      - ReadWriteMany
      resources:
        requests:
          storage: 1Mi
      volumeName: existing-pv-name-goes-here
      storageClassName: ""
  EOF
  )
}
```

**Comment by VardyNg** [comment]
+1 to this issue. 

I am configuring a AWS S3 volume mount for EKS, and found the PVC's `storage_class_name = ""` being treated as `storage_class_name = nil`, the PVC storage class became `gp2` (the default class). And it failed to attach to the PV because their `storage_class_name` does not align.

```terraform
resource "kubernetes_persistent_volume" "default" {
  metadata {
    name = "${local.name}-pv"
  }
  spec {
    capacity = {
      storage = "5Gi"
    }
    access_modes = ["ReadWriteMany"]
    storage_class_name = "" # it is treated as empty string correctly
    persistent_volume_source {
      csi {
        driver = "s3.csi.aws.com"
        volume_handle = "s3-csi-driver-volume"
        volume_attributes = {
          bucketName: aws_s3_bucket.default.id
        }
      }
    }
  }
}

resource "kubernetes_persistent_volume_claim" "default" {
  metadata {
    name = "${local.name}-pvc"
    namespace = kubernetes_namespace_v1.ns.metadata.0.name
  }
  spec {
    access_modes = ["ReadWriteMany"]
    storage_class_name = "" # <- being treated as nil
    resources {
      requests = {
        storage = "3Gi"
      }
    }
    volume_name = kubernetes_persistent_volume.default.metadata.0.name
  }
  wait_until_bound = false
}
```

**Comment by Lillecarl** [comment]
@VardyNg use kubernetes_manifest instead, it just works ™

**Comment by VardyNg** [comment]
> @VardyNg use kubernetes_manifest instead, it just works ™

It works! Hopefully there will be a fix for this parameter as well

**Comment by marek-knappe** [comment]
Still not fixed :/ will switch to manifest

**Comment by Lillecarl** [comment]
@marek-knappe https://registry.terraform.io/providers/alekc/kubectl/latest/docs/resources/kubectl_manifest

**Comment by nmusatti** [comment]
This is still broken in version 2.36.0.
